Add saving email specific styles [MAILPOET-5640]

lysyjan opened the pull request:
## Description

_N/A_

## Code review notes

This is the final pull request related to the layout styles.
- It adds saving post metadata such as styles.
- It activates block-editor private API.
- It creates a new component that renders styles and applies changes via the styles panel.

## QA notes

_N/A_

## Linked PRs

- https://github.com/mailpoet/mailpoet/pull/5491
- https://github.com/mailpoet/mailpoet/pull/5497

## Linked tickets

[MAILPOET-5640]

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5640]: https://mailpoet.atlassian.net/browse/MAILPOET-5640?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Add theme metadata to mailpoet_email post

[MAILPOET-5640]

lysyjan committed:
Pass email editor theme to the state

[MAILPOET-5640]

lysyjan committed:
Add private-api package

[MAILPOET-5640]

lysyjan committed:
Add theme styles component for email editor

[MAILPOET-5640]

lysyjan committed:
Update email theme via dimensions-panel

[MAILPOET-5640]

lysyjan committed:
Remove redundant code

[MAILPOET-5640]

lysyjan committed:
Duplicate email post meta and update tests

[MAILPOET-5640]

lysyjan committed:
Update integration tests

[MAILPOET-5640]

lysyjan committed:
Use VariablesPostprocessor in Renderer

Because the layout wrapper HTML is not post-processed after moving content rendering, I needed to use VariablesPostprocessor because padding can be configured by a CSS variable.
[MAILPOET-5640]

costasovo commented on mailpoet/assets/js/src/email-editor/engine/components/block-editor/block-editor.tsx:
```diff
@@ -14,6 +14,9 @@ import {
   // @ts-ignore No types for this exist yet.
   __unstableEditorStyles as EditorStyles,
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore No types for this exist yet.
```
Could we use ts-expect-error here?

lysyjan replied:
👍 I tried to address this across the whole email editor: https://github.com/mailpoet/mailpoet/pull/5499/commits/ba5d61f6325193328c907070b2c138f4719bee89

costasovo commented on mailpoet/assets/js/src/email-editor/engine/components/block-editor/block-editor.tsx:
```diff
@@ -122,82 +125,92 @@ export function BlockEditor() {
+    <BlockContextProvider
+      value={{
+        postId,
+        postType: 'mailpoet_email',
+        templateSlug: 'mailpoet-email',
```
Just a detail. It looks a bit weird to see `mailpoet_email` and `mailpoet-email`. Could we maybe use `mailpoet_email_template`? If I understand correctly it doesn't matter what we use, right?

lysyjan replied:
👍 Yes, it doesn't matter about the name. Changed here: https://github.com/mailpoet/mailpoet/pull/5499/commits/c7ad438e5e703a286a60154eb0d081764459a242

costasovo commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/EmailEditor.php:
```diff
@@ -51,6 +55,17 @@ public function addEmailPostType(array $postTypes): array {
+            'show_in_rest' => true,
+            'single' => true,
+            'type' => 'object',
+            'default' => ['version' => 2],
```
I wonder why we use `['version' => 2]`?

lysyjan replied:
If we don't use the same version as it is in our theme.json files, the merge doesn't work properly. I added a comment.

costasovo commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/EmailEditor.php:
```diff
@@ -85,4 +100,12 @@ public function getEmailDefaultContent(): string {
+      $theme->merge(new WP_Theme_JSON($emailTheme));
+    }
+    return $theme;
+  }
```
I think that saving/fetching and the definition of the meta key for the styles settings should be part of the engine rather than the MailPoet integration. I mean the editor needs them so we shouldn't rely on an integration to provide those.

lysyjan replied:
Good point. Moved to the engine: https://github.com/mailpoet/mailpoet/pull/5499/commits/642f5f5cf4361db2529620b0cea38327fa6704d1

costasovo commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/EmailApiController.php:
```diff
@@ -49,6 +50,8 @@ public function saveEmailData(array $data, \WP_Post $emailPost): void {
     }
 
+    update_post_meta($emailPost->ID, EmailEditor::MAILPOET_EMAIL_META_THEME_TYPE, $data['theme']);
+
```
I think we are mixing two concepts here.

I like the idea of saving the theme in post-meta.  We are using [register_rest_field](https://developer.wordpress.org/reference/functions/register_rest_field/) for saving/fetching some data we store in custom tables.

But for post meta, I would expect the WP Rest API to have built-in handling so we don't have to pass them around via our extra field, but we should be able to fetch them and update them directly via core/store.
I googled a bit and found this post https://jeffreycarandang.com/how-to-enable-read-and-update-post-meta-in-the-wordpress-block-editor-gutenberg/

So could we use the built-in API for fetching/saving post meta fields. That way we can move it to the engine more easily.

lysyjan replied:
@costasovo Great point here! I refactored this a bit, and the final code looks better. Please check all new commits.

costasovo reviewed (commented):
@lysyjan Good job!

I added a couple of comments with questions and suggestions. I hope they make sense.

lysyjan committed:
Replace @ts-ignore with @ts-expects errors in the email editor

[MAILPOET-5640]

lysyjan committed:
Change template slug in block-editor

[MAILPOET-5640]

lysyjan committed:
Move meta registration to the engine

As a part of the refactoring is making the meta field available via API.
[MAILPOET-5640]

lysyjan committed:
Move email styles merge to engine

[MAILPOET-5640]

lysyjan committed:
Remove redundant property theme

Because we can write and read meta field directly, we don't need our own implementation.
[MAILPOET-5640]

costasovo reviewed (dismissed):
Thank you for fixing the usage of metadata! 👍 LGTM!

lysyjan committed:
Fix ordering used classes after rebase

[MAILPOET-5640]
